### ISSUE TITLE
UniqueIndex does not prevent adding an element...

### ISSUE THREAD

**Issue by victornoel** [npgall/cqengine#177]
I am using a `ConcurrentIndexedCollection`, when I add an element that violates an `UniqueIndex`, I get an exception, but the element is still added to the collection, is that expected?

Looking at the code, https://github.com/npgall/cqengine/blob/2.12.4/code/src/main/java/com/googlecode/cqengine/ConcurrentIndexedCollection.java#L350, it is visible that the element is added before indexes are checked.

**Comment by npgall** [comment]
Hi victornoel,

Yes that is expected actually. It's mentioned in the [JavaDocs](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/index/unique/UniqueIndex.html): 

> "This index will throw an exception if a duplicate object is detected for an existing attribute value. That condition means however that inconsistencies might already have arisen between this and other indexes as a result of the application's misuse of this index."

Basically, the intent of the `UniqueIndex` is not to enforce a unique constraint. It is to index attributes which are already guaranteed by the application to uniquely identify an object (primary key-type attributes etc.). Often an external system such as a database from which the objects originate might inherently ensure uniqueness, so CQEngine doesn't attempt to re-verify this. It is up to the developer to ensure that `UniqueIndex` is only used with values which are unique.

My reasoning for this approach is that having to check with each index in advance if it _would_ accept an object, before we actually insert the object, would introduce more overhead in the write path. So most indexes are trusted to simply accept new objects. It's only in exceptional circumstances that an index would not accept an object - such as when some requirement of the index is not met such as this.

The purpose of `UniqueIndex`is to avoid allocating bucket data structures for holding multiple values, which can improve efficiency say compared to the `HashIndex`. However for applications which cannot always guarantee uniqueness, there is also [HashIndex.onSemiUniqueAttribute()](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/index/hash/HashIndex.html#onSemiUniqueAttribute-com.googlecode.cqengine.attribute.Attribute-) which will reduce the bucket size in `HashIndex` as a middle ground.

Hope that helps!

**Comment by victornoel** [comment]
Ok, that makes sense, I misunderstood the responsibility of the indexes then :)

**Comment by victornoel** [comment]
(thanks a lot for the the lengthy explanation!)

**Comment by victornoel** [comment]
@npgall still, I forgot about that point, but it is too bad that inserting an element (that violates the `UniqueIndex`) into the collection leaves it in a incorrect state: because the `UniqueIndex` throw the exception, all the indexes that were supposed to be executed after are thus not executed.
You end up with a collection containing the element and half of the indexes populated regarding this element.

What would be the recommended way to handle such situation or what could be done to improve cqengine on that front?

**Comment by victornoel** [comment]
(or is that a use case that is not meant to be covered by cqengine?)

for the record, I am not blocked by that situation, just wondering :)

**Comment by npgall** [comment]
Sorry for the delay replying!

The best workaround if your application cannot guarantee the uniqueness of values for an attribute is to use a `HashIndex` instead - possibly via [HashIndex.onSemiUniqueAttribute()](http://htmlpreview.github.io/?http://raw.githubusercontent.com/npgall/cqengine/master/documentation/javadoc/apidocs/com/googlecode/cqengine/index/hash/HashIndex.html#onSemiUniqueAttribute-com.googlecode.cqengine.attribute.Attribute-) if the values could at least be considered _semi-unique_.

The only reason for `UniqueIndex` to exist is to leverage the uniqueness of values to improve performance. It doesn't provide any features that `HashIndex` doesn't already provide. So as the only reason to use it is performance, it doesn't attempt to check for constraint violations in a more sophisticated way, because doing so would harm performance in applications where the values actually are guaranteed to be unique. It's basically a design tradeoff.

I guess another question then is _how to apply constraints_ on the data added to an IndexedCollection? I think this is an area which could be improved, and it might be easy enough to do so in `TransactionalIndexedCollection`. Feel free to open another issue called "Improve support for constraints" or similar and I will try to look at that when I get a chance.

**Comment by victornoel** [comment]
@npgall thanks for the explanation!

my problem right now is more about the fact that there is a potential of leaving the collection in an incoherent state. UniqueIndex should maybe not be throwing an exception, or maybe the exception of all the indexes should be grouped using `addSuppressed` of a main exception which will be thrown only after all indexes did their job?

I don't know if I'm clear ^^

**Comment by DominicAnelloSonos** [comment]
@victornoel - I found this issue because I had the exact same problem.  In the end I ended up putting a guard around my add element logic that dynamically checks all the `UniqueIndex`s against the incoming value and raises an exception if a duplicate is found before adding the new element.  This is probably massively inefficient, but since I am using CQEngine in a mock for unit testing and only handling a few hundred values at most it is acceptable to me.

Dropping it here in case it is useful to someone else.

```java
class CqengineRepoistory<T> {
    private final IndexedCollection<T> elements = new ObjectLockingIndexedCollection<>();

    @SuppressWarnings({ "unchecked", "rawtypes" })
    public void addElement(T newElement) {
        StreamSupport.stream(this.elements.getIndexes().spliterator(), false)
            .filter(x -> x instanceof UniqueIndex)
            .map(x -> (UniqueIndex) x)
            .forEach(x -> {
                final Attribute attr = x.getAttribute();
                attr.getValues(u, null).forEach(incomingValue -> {
                    ResultSet<T> rs = x.retrieve(QueryFactory.equal(attr, incomingValue), null);
                    if (rs.isNotEmpty()) {
                        throw new UniqueContraintViolationException("UniqueIndex on attribute `" 
                            + attr.getAttributeName() + "` violated. Problematic attribute value: '" 
                            + incomingValue + "', problematic duplicate object: " + newElement);
                    }
                });
            });

        elements.add(newElement);
    }
```

**Comment by victornoel** [comment]
@DominicAnelloSonos thx for the idea :)
